odoo/addons/base/models/assetsbundle.py

```diff
@@ -322,13 +342,71 @@ def save_attachment(self, type, content):
 
         return attachment
 
-    def js(self):
-        attachments = self.get_attachments('js')
-        if not attachments:
-            content = ';\n'.join(asset.minify() for asset in self.javascripts)
-            return self.save_attachment('js', content)
+    def js(self, is_minified=True, with_sourcemap=False, name_unique=True):
+        url = None if name_unique else "/%s.js" % self.name
+        attachments = self.get_attachments('js', is_minified=is_minified, url_pattern=url)
+        if not attachments or (not name_unique and bundle_checksum.get(self.name, None) != self.checksum):
+            if not name_unique:
+                bundle_checksum[self.name] = self.checksum
+            if (with_sourcemap):
+                return self.js_with_sourcemap(name_unique=name_unique)
+            else:
+                content = ';\n'.join(asset.minify() if is_minified else asset.with_header(asset.content, minimal=False)
+                                    for asset in self.javascripts)
+                return self.save_attachment('js', content, is_minified=is_minified, url=url)
+
         return attachments[0]
 
+    def js_with_sourcemap(self, name_unique=True):
+        content_bundle = ""
+        generator = SourceMapGenerator({
+            "file": "/%s.js" % self.name,
+            "sourceRoot": "./"
+        })
+        content_line_count = 0
+        line_header = 6
+        for asset in self.javascripts:
+            if asset.is_odoo_module:
+                generator.add_source(
+                    asset.url, asset._content, content_line_count, nb_skip_line_output=line_header+3)
+                asset_line_count = len(asset.content.split("\n")) + 6
+            else:
+                asset_line_count = generator.add_source(
+                    asset.url, asset.content, content_line_count, nb_skip_line_output=line_header)
+
+            if (content_bundle == ""):
+                content_bundle = asset.with_header(asset.content, minimal=False)
+            else:
+                content_bundle += ';\n' + asset.with_header(asset.content, minimal=False)
```

Comment:
Such string concatenations do not scale well (every concatenation allocates and copies data in memory), and the condition for when to add a semicolon and line break is ugly.  The more common idiom is to put all pieces in a list, and then join them with a separator:
```python
content_list = []
for asset in self.javascripts:
    ...
    content_list.append(asset.with_header(asset.content, minimal=False))

content_bundle = ";\n".join(content_list)
asset_line_count = len(content_bundle.splitlines())    # computing on final result is safer
```

Reply:
Beware this is not strictly equivalent to the above, there is a missing ";\n" at the end of the joined string.

Reply:
Shouldn't the last line ends with a ; too ?

Reply:
> Such string concatenations do not scale well (every concatenation allocates and copies data in memory)

TBF that's very hard to trigger in CPython because [it "cheats" and updates the LHS in-place if that only has one owner](https://docs.python.org/2/whatsnew/2.4.html#optimizations). It will only go superlinear once `realloc` starts moving strings on every iteration, which will depend on the platform but IIRC happens in the megabyte range.

Obviously this doesn't apply to third-party implementations like pypy which can't play around with a refcount and will go quadratic immediately.

Reply:
asset_line_count must be calculated for each loop round for the sourcemap generator. I'm ok to remove the concatenation